# Language Translation
In this project, you’re going to take a peek into the realm of neural network machine translation.  You’ll be training a sequence to sequence model on a dataset of English and French sentences that can translate new sentences from English to French.
## Get the Data
Since translating the whole language of English to French will take lots of time to train, we have provided you with a small portion of the English corpus.

In [4]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
import helper
import problem_unittests as tests

source_path = 'data/small_vocab_en'
target_path = 'data/small_vocab_fr'
source_text = helper.load_data(source_path)
target_text = helper.load_data(target_path)

## Explore the Data
Play around with view_sentence_range to view different parts of the data.

In [2]:
view_sentence_range = (0, 10)

"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
import numpy as np

print('Dataset Stats')
print('Roughly the number of unique words: {}'.format(len({word: None for word in source_text.split()})))

sentences = source_text.split('\n')
word_counts = [len(sentence.split()) for sentence in sentences]
print('Number of sentences: {}'.format(len(sentences)))
print('Average number of words in a sentence: {}'.format(np.average(word_counts)))

print()
print('English sentences {} to {}:'.format(*view_sentence_range))
print('\n'.join(source_text.split('\n')[view_sentence_range[0]:view_sentence_range[1]]))
print()
print('French sentences {} to {}:'.format(*view_sentence_range))
print('\n'.join(target_text.split('\n')[view_sentence_range[0]:view_sentence_range[1]]))

Dataset Stats
Roughly the number of unique words: 227
Number of sentences: 137861
Average number of words in a sentence: 13.225277634719028

English sentences 0 to 10:
new jersey is sometimes quiet during autumn , and it is snowy in april .
the united states is usually chilly during july , and it is usually freezing in november .
california is usually quiet during march , and it is usually hot in june .
the united states is sometimes mild during june , and it is cold in september .
your least liked fruit is the grape , but my least liked is the apple .
his favorite fruit is the orange , but my favorite is the grape .
paris is relaxing during december , but it is usually chilly in july .
new jersey is busy during spring , and it is never hot in march .
our least liked fruit is the lemon , but my least liked is the grape .
the united states is sometimes busy during january , and it is sometimes warm in november .

French sentences 0 to 10:
new jersey est parfois calme pendant l' automne 

## Implement Preprocessing Function
### Text to Word Ids
As you did with other RNNs, you must turn the text into a number so the computer can understand it. In the function `text_to_ids()`, you'll turn `source_text` and `target_text` from words to ids.  However, you need to add the `<EOS>` word id at the end of `target_text`.  This will help the neural network predict when the sentence should end.

You can get the `<EOS>` word id by doing:
```python
target_vocab_to_int['<EOS>']
```
You can get other word ids using `source_vocab_to_int` and `target_vocab_to_int`.

In [34]:
def text_to_ids(source_text, target_text, source_vocab_to_int, target_vocab_to_int):
    """
    Convert source and target text to proper word ids
    :param source_text: String that contains all the source text.
    :param target_text: String that contains all the target text.
    :param source_vocab_to_int: Dictionary to go from the source words to an id
    :param target_vocab_to_int: Dictionary to go from the target words to an id
    :return: A tuple of lists (source_id_text, target_id_text)
    """
    source_id_text = []
    sentences_source = source_text.split('\n')
    for sentence in sentences_source:
        sentence_ids = []
        for word in sentence.split():
            sentence_ids.append(source_vocab_to_int[word])
        source_id_text.append(sentence_ids)

    target_id_text = []
    sentences_target = target_text.split('\n')
    for sentence in sentences_target:
        sentence_ids = []
        for word in sentence.split():
            sentence_ids.append(target_vocab_to_int[word])
        sentence_ids.append(target_vocab_to_int['<EOS>'])
        target_id_text.append(sentence_ids)
    
    return source_id_text, target_id_text

"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_text_to_ids(text_to_ids)

Tests Passed


### Preprocess all the data and save it
Running the code cell below will preprocess all the data and save it to file.

In [35]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
helper.preprocess_and_save_data(source_path, target_path, text_to_ids)

# Check Point
This is your first checkpoint. If you ever decide to come back to this notebook or have to restart the notebook, you can start from here. The preprocessed data has been saved to disk.

In [1]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
import numpy as np
import helper
import problem_unittests as tests


(source_int_text, target_int_text), (source_vocab_to_int, target_vocab_to_int), _ = helper.load_preprocess()

### Check the Version of TensorFlow and Access to GPU
This will check to make sure you have the correct version of TensorFlow and access to a GPU

In [2]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
from distutils.version import LooseVersion
import warnings
import tensorflow as tf
from tensorflow.python.layers.core import Dense

# Check TensorFlow Version
assert LooseVersion(tf.__version__) >= LooseVersion('1.1'), 'Please use TensorFlow version 1.1 or newer'
print('TensorFlow Version: {}'.format(tf.__version__))

# Check for a GPU
if not tf.test.gpu_device_name():
    warnings.warn('No GPU found. Please use a GPU to train your neural network.')
else:
    print('Default GPU Device: {}'.format(tf.test.gpu_device_name()))

TensorFlow Version: 1.2.1
Default GPU Device: /gpu:0


## Build the Neural Network
You'll build the components necessary to build a Sequence-to-Sequence model by implementing the following functions below:
- `model_inputs`
- `process_decoder_input`
- `encoding_layer`
- `decoding_layer_train`
- `decoding_layer_infer`
- `decoding_layer`
- `seq2seq_model`

### Input
Implement the `model_inputs()` function to create TF Placeholders for the Neural Network. It should create the following placeholders:

- Input text placeholder named "input" using the TF Placeholder name parameter with rank 2.
- Targets placeholder with rank 2.
- Learning rate placeholder with rank 0.
- Keep probability placeholder named "keep_prob" using the TF Placeholder name parameter with rank 0.
- Target sequence length placeholder named "target_sequence_length" with rank 1
- Max target sequence length tensor named "max_target_len" getting its value from applying tf.reduce_max on the target_sequence_length placeholder. Rank 0.
- Source sequence length placeholder named "source_sequence_length" with rank 1

Return the placeholders in the following the tuple (input, targets, learning rate, keep probability, target sequence length, max target sequence length, source sequence length)

In [3]:
def model_inputs():
    """
    Create TF Placeholders for input, targets, learning rate, and lengths of source and target sequences.
    :return: Tuple (input, targets, learning rate, keep probability, target sequence length,
    max target sequence length, source sequence length)
    """
    # TODO: Implement Function
    target_sequence_length = tf.placeholder(tf.int32, shape = [None], name = "target_sequence_length")

    return tf.placeholder(tf.int32,shape=[None, None], name = "input"), \
           tf.placeholder(tf.int32, shape = [None, None]), \
           tf.placeholder(tf.float32, shape = []), \
           tf.placeholder(tf.float32, shape = [], name = "keep_prob"), \
           target_sequence_length,\
           tf.reduce_max(target_sequence_length, name = "max_target_len"), \
           tf.placeholder(tf.int32, shape = [None], name = "source_sequence_length")


"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_model_inputs(model_inputs)

Tests Passed


### Process Decoder Input
Implement `process_decoder_input` by removing the last word id from each batch in `target_data` and concat the GO ID to the begining of each batch.

In [4]:
def process_decoder_input(target_data, target_vocab_to_int, batch_size):
    """
    Preprocess target data for encoding
    :param target_data: Target Placehoder
    :param target_vocab_to_int: Dictionary to go from the target words to an id
    :param batch_size: Batch Size
    :return: Preprocessed target data
    """
    ending = tf.strided_slice(target_data, [0, 0], [batch_size, -1], [1, 1])
    dec_input = tf.concat([tf.fill([batch_size, 1], target_vocab_to_int['<GO>']), ending], 1)

    return dec_input

"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_process_encoding_input(process_decoder_input)

Tests Passed


### Encoding
Implement `encoding_layer()` to create a Encoder RNN layer:
 * Embed the encoder input using [`tf.contrib.layers.embed_sequence`](https://www.tensorflow.org/api_docs/python/tf/contrib/layers/embed_sequence)
 * Construct a [stacked](https://github.com/tensorflow/tensorflow/blob/6947f65a374ebf29e74bb71e36fd82760056d82c/tensorflow/docs_src/tutorials/recurrent.md#stacking-multiple-lstms) [`tf.contrib.rnn.LSTMCell`](https://www.tensorflow.org/api_docs/python/tf/contrib/rnn/LSTMCell) wrapped in a [`tf.contrib.rnn.DropoutWrapper`](https://www.tensorflow.org/api_docs/python/tf/contrib/rnn/DropoutWrapper)
 * Pass cell and embedded input to [`tf.nn.dynamic_rnn()`](https://www.tensorflow.org/api_docs/python/tf/nn/dynamic_rnn)

In [5]:
from imp import reload
reload(tests)

def make_cell(rnn_size, keep_prob):
    cell = tf.contrib.rnn.DropoutWrapper(tf.contrib.rnn.LSTMCell(rnn_size,
                                   initializer=tf.random_uniform_initializer(-0.1, 0.1, seed=2)),
                                    output_keep_prob=keep_prob)
    return cell

def encoding_layer(rnn_inputs, rnn_size, num_layers, keep_prob, 
                   source_sequence_length, source_vocab_size, 
                   encoding_embedding_size):
    """
    Create encoding layer
    :param rnn_inputs: Inputs for the RNN
    :param rnn_size: RNN Size
    :param num_layers: Number of layers
    :param keep_prob: Dropout keep probability
    :param source_sequence_length: a list of the lengths of each sequence in the batch
    :param source_vocab_size: vocabulary size of source data
    :param encoding_embedding_size: embedding size of source data
    :return: tuple (RNN output, RNN state)
    """
    emb = tf.contrib.layers.embed_sequence(
        rnn_inputs,
        vocab_size=source_vocab_size,
        embed_dim=encoding_embedding_size,
    )
    


    enc_cell = tf.contrib.rnn.MultiRNNCell([make_cell(rnn_size, keep_prob) for _ in range(num_layers)])

    
    enc_output, enc_state = tf.nn.dynamic_rnn(enc_cell, emb, sequence_length=source_sequence_length, dtype=tf.float32)

    return enc_output, enc_state

"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_encoding_layer(encoding_layer)

Tests Passed


### Decoding - Training
Create a training decoding layer:
* Create a [`tf.contrib.seq2seq.TrainingHelper`](https://www.tensorflow.org/api_docs/python/tf/contrib/seq2seq/TrainingHelper) 
* Create a [`tf.contrib.seq2seq.BasicDecoder`](https://www.tensorflow.org/api_docs/python/tf/contrib/seq2seq/BasicDecoder)
* Obtain the decoder outputs from [`tf.contrib.seq2seq.dynamic_decode`](https://www.tensorflow.org/api_docs/python/tf/contrib/seq2seq/dynamic_decode)

In [6]:

def decoding_layer_train(encoder_state, dec_cell, dec_embed_input, 
                         target_sequence_length, max_summary_length, 
                         output_layer, keep_prob):
    """
    Create a decoding layer for training
    :param encoder_state: Encoder State OK
    :param dec_cell: Decoder RNN Cell OK
    :param dec_embed_input: Decoder embedded input OK
    :param target_sequence_length: The lengths of each sequence in the target batch OK
    :param max_summary_length: The length of the longest sequence in the batch OK
    :param output_layer: Function to apply the output layer OK
    :param keep_prob: Dropout keep probability OK
    :return: BasicDecoderOutput containing training logits and sample_id
    """
    train_helper = tf.contrib.seq2seq.TrainingHelper(dec_embed_input, target_sequence_length)
    basic_decoder = tf.contrib.seq2seq.BasicDecoder(tf.contrib.rnn.DropoutWrapper(dec_cell, input_keep_prob=keep_prob), train_helper, encoder_state, output_layer)
    final_outputs, _, _ =  tf.contrib.seq2seq.dynamic_decode(basic_decoder, maximum_iterations=max_summary_length)

    return final_outputs
    



"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_decoding_layer_train(decoding_layer_train)

Tests Passed


### Decoding - Inference
Create inference decoder:
* Create a [`tf.contrib.seq2seq.GreedyEmbeddingHelper`](https://www.tensorflow.org/api_docs/python/tf/contrib/seq2seq/GreedyEmbeddingHelper)
* Create a [`tf.contrib.seq2seq.BasicDecoder`](https://www.tensorflow.org/api_docs/python/tf/contrib/seq2seq/BasicDecoder)
* Obtain the decoder outputs from [`tf.contrib.seq2seq.dynamic_decode`](https://www.tensorflow.org/api_docs/python/tf/contrib/seq2seq/dynamic_decode)

In [7]:
def decoding_layer_infer(encoder_state, dec_cell, dec_embeddings, start_of_sequence_id,
                         end_of_sequence_id, max_target_sequence_length,
                         vocab_size, output_layer, batch_size, keep_prob):
    """
    Create a decoding layer for inference
    :param encoder_state: Encoder state
    :param dec_cell: Decoder RNN Cell
    :param dec_embeddings: Decoder embeddings
    :param start_of_sequence_id: GO ID
    :param end_of_sequence_id: EOS Id
    :param max_target_sequence_length: Maximum length of target sequences
    :param vocab_size: Size of decoder/target vocabulary
    :param decoding_scope: TenorFlow Variable Scope for decoding 
    :param output_layer: Function to apply the output layer
    :param batch_size: Batch size
    :param keep_prob: Dropout keep probability
    :return: BasicDecoderOutput containing inference logits and sample_id
    """
    start_tokens = tf.tile(tf.constant([start_of_sequence_id], dtype=tf.int32), [batch_size], name='start_tokens')
    inference_helper = tf.contrib.seq2seq.GreedyEmbeddingHelper(dec_embeddings,
                                                                start_tokens,
                                                                end_of_sequence_id)
    inference_decoder = tf.contrib.seq2seq.BasicDecoder(dec_cell,
                                                        inference_helper,
                                                        encoder_state,
                                                        output_layer)
    
    inference_decoder_output, _, _ = tf.contrib.seq2seq.dynamic_decode(inference_decoder,
                                                            impute_finished=True,
                                                            maximum_iterations=max_target_sequence_length)
        
        
    


    return inference_decoder_output



"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_decoding_layer_infer(decoding_layer_infer)

Tests Passed


### Build the Decoding Layer
Implement `decoding_layer()` to create a Decoder RNN layer.

* Embed the target sequences
* Construct the decoder LSTM cell (just like you constructed the encoder cell above)
* Create an output layer to map the outputs of the decoder to the elements of our vocabulary
* Use the your `decoding_layer_train(encoder_state, dec_cell, dec_embed_input, target_sequence_length, max_target_sequence_length, output_layer, keep_prob)` function to get the training logits.
* Use your `decoding_layer_infer(encoder_state, dec_cell, dec_embeddings, start_of_sequence_id, end_of_sequence_id, max_target_sequence_length, vocab_size, output_layer, batch_size, keep_prob)` function to get the inference logits.

Note: You'll need to use [tf.variable_scope](https://www.tensorflow.org/api_docs/python/tf/variable_scope) to share variables between training and inference.

In [8]:
def decoding_layer(dec_input, encoder_state,
                   target_sequence_length, max_target_sequence_length,
                   rnn_size,
                   num_layers, target_vocab_to_int, target_vocab_size,
                   batch_size, keep_prob, decoding_embedding_size):
    """
    Create decoding layer
    :param dec_input: Decoder input
    :param encoder_state: Encoder state
    :param target_sequence_length: The lengths of each sequence in the target batch
    :param max_target_sequence_length: Maximum length of target sequences
    :param rnn_size: RNN Size
    :param num_layers: Number of layers
    :param target_vocab_to_int: Dictionary to go from the target words to an id
    :param target_vocab_size: Size of target vocabulary
    :param batch_size: The size of the batch
    :param keep_prob: Dropout keep probability
    :param decoding_embedding_size: Decoding embedding size
    :return: Tuple of (Training BasicDecoderOutput, Inference BasicDecoderOutput)
    """
    dec_embeddings = tf.Variable(tf.random_uniform([target_vocab_size, decoding_embedding_size]))
    dec_embed_input = tf.nn.embedding_lookup(dec_embeddings, dec_input)

    # 2. Construct the decoder cell
    def make_cell(rnn_size):
        dec_cell = tf.contrib.rnn.LSTMCell(rnn_size,
                                           initializer=tf.random_uniform_initializer(-0.1, 0.1, seed=2))
        return dec_cell

    dec_cell = tf.contrib.rnn.MultiRNNCell([make_cell(rnn_size) for _ in range(num_layers)])
     
    # 3. Dense layer to translate the decoder's output at each time 
    # step into a choice from the target vocabulary
    output_layer = Dense(target_vocab_size,
                         kernel_initializer = tf.truncated_normal_initializer(mean = 0.0, stddev=0.1))

    
    
    with tf.variable_scope("decode"):
        training_decoder_output = decoding_layer_train(
            encoder_state,
            dec_cell,
            dec_embed_input,
            target_sequence_length,
            max_target_sequence_length,
            output_layer, keep_prob)
    
    with tf.variable_scope("decode", reuse=True):
        inference_decoder_output = decoding_layer_infer(
            encoder_state,
            dec_cell,
            dec_embeddings,
            target_vocab_to_int['<GO>'],
            target_vocab_to_int['<EOS>'],
            max_target_sequence_length,
            target_vocab_size,
            output_layer,
            batch_size, keep_prob)

        
        
    return training_decoder_output, inference_decoder_output



"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_decoding_layer(decoding_layer)

Tests Passed


### Build the Neural Network
Apply the functions you implemented above to:

- Encode the input using your `encoding_layer(rnn_inputs, rnn_size, num_layers, keep_prob,  source_sequence_length, source_vocab_size, encoding_embedding_size)`.
- Process target data using your `process_decoder_input(target_data, target_vocab_to_int, batch_size)` function.
- Decode the encoded input using your `decoding_layer(dec_input, enc_state, target_sequence_length, max_target_sentence_length, rnn_size, num_layers, target_vocab_to_int, target_vocab_size, batch_size, keep_prob, dec_embedding_size)` function.

In [9]:
def seq2seq_model(input_data, target_data, keep_prob, batch_size,
                  source_sequence_length, target_sequence_length,
                  max_target_sentence_length,
                  source_vocab_size, target_vocab_size,
                  enc_embedding_size, dec_embedding_size,
                  rnn_size, num_layers, target_vocab_to_int):
    """
    Build the Sequence-to-Sequence part of the neural network
    :param input_data: Input placeholder
    :param target_data: Target placeholder
    :param keep_prob: Dropout keep probability placeholder
    :param batch_size: Batch Size
    :param source_sequence_length: Sequence Lengths of source sequences in the batch
    :param target_sequence_length: Sequence Lengths of target sequences in the batch
    :param source_vocab_size: Source vocabulary size
    :param target_vocab_size: Target vocabulary size
    :param enc_embedding_size: Decoder embedding size
    :param dec_embedding_size: Encoder embedding size
    :param rnn_size: RNN Size
    :param num_layers: Number of layers
    :param target_vocab_to_int: Dictionary to go from the target words to an id
    :return: Tuple of (Training BasicDecoderOutput, Inference BasicDecoderOutput)
    """
    _, enc_state = encoding_layer(input_data,
                                  rnn_size,
                                  num_layers,
                                  keep_prob,
                                  source_sequence_length,
                                  source_vocab_size,
                                  enc_embedding_size)
    
    dec_input =  process_decoder_input(target_data, target_vocab_to_int, batch_size)
    
    training_decoder_output, inference_decoder_output = decoding_layer(dec_input,
                                                                       enc_state,
                                                                       target_sequence_length,
                                                                       max_target_sentence_length,
                                                                       rnn_size,
                                                                       num_layers,
                                                                       target_vocab_to_int,
                                                                       target_vocab_size,
                                                                       batch_size,
                                                                       keep_prob,
                                                                       dec_embedding_size)

    
    
    
    
    return training_decoder_output, inference_decoder_output


"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_seq2seq_model(seq2seq_model)

Tests Passed


## Neural Network Training
### Hyperparameters
Tune the following parameters:

- Set `epochs` to the number of epochs.
- Set `batch_size` to the batch size.
- Set `rnn_size` to the size of the RNNs.
- Set `num_layers` to the number of layers.
- Set `encoding_embedding_size` to the size of the embedding for the encoder.
- Set `decoding_embedding_size` to the size of the embedding for the decoder.
- Set `learning_rate` to the learning rate.
- Set `keep_probability` to the Dropout keep probability
- Set `display_step` to state how many steps between each debug output statement

In [25]:
# Number of Epochs
epochs = 8
# Batch Size
batch_size = 256
# RNN Size
rnn_size = 256
# Number of Layers
num_layers = 2
# Embedding Size
encoding_embedding_size = 128
decoding_embedding_size = 128
# Learning Rate
learning_rate = 0.001
# Dropout Keep Probability
keep_probability = 0.7
display_step = True

### Build the Graph
Build the graph using the neural network you implemented.

In [26]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
save_path = 'checkpoints/dev'
(source_int_text, target_int_text), (source_vocab_to_int, target_vocab_to_int), _ = helper.load_preprocess()
max_target_sentence_length = max([len(sentence) for sentence in source_int_text])

train_graph = tf.Graph()
with train_graph.as_default():
    input_data, targets, lr, keep_prob, target_sequence_length, max_target_sequence_length, source_sequence_length = model_inputs()

    #sequence_length = tf.placeholder_with_default(max_target_sentence_length, None, name='sequence_length')
    input_shape = tf.shape(input_data)

    train_logits, inference_logits = seq2seq_model(tf.reverse(input_data, [-1]),
                                                   targets,
                                                   keep_prob,
                                                   batch_size,
                                                   source_sequence_length,
                                                   target_sequence_length,
                                                   max_target_sequence_length,
                                                   len(source_vocab_to_int),
                                                   len(target_vocab_to_int),
                                                   encoding_embedding_size,
                                                   decoding_embedding_size,
                                                   rnn_size,
                                                   num_layers,
                                                   target_vocab_to_int)


    training_logits = tf.identity(train_logits.rnn_output, name='logits')
    inference_logits = tf.identity(inference_logits.sample_id, name='predictions')

    masks = tf.sequence_mask(target_sequence_length, max_target_sequence_length, dtype=tf.float32, name='masks')

    with tf.name_scope("optimization"):
        # Loss function
        cost = tf.contrib.seq2seq.sequence_loss(
            training_logits,
            targets,
            masks)

        # Optimizer
        optimizer = tf.train.AdamOptimizer(lr)

        # Gradient Clipping
        gradients = optimizer.compute_gradients(cost)
        capped_gradients = [(tf.clip_by_value(grad, -1., 1.), var) for grad, var in gradients if grad is not None]
        train_op = optimizer.apply_gradients(capped_gradients)


Batch and pad the source and target sequences

In [27]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
def pad_sentence_batch(sentence_batch, pad_int):
    """Pad sentences with <PAD> so that each sentence of a batch has the same length"""
    max_sentence = max([len(sentence) for sentence in sentence_batch])
    return [sentence + [pad_int] * (max_sentence - len(sentence)) for sentence in sentence_batch]


def get_batches(sources, targets, batch_size, source_pad_int, target_pad_int):
    """Batch targets, sources, and the lengths of their sentences together"""
    for batch_i in range(0, len(sources)//batch_size):
        start_i = batch_i * batch_size

        # Slice the right amount for the batch
        sources_batch = sources[start_i:start_i + batch_size]
        targets_batch = targets[start_i:start_i + batch_size]

        # Pad
        pad_sources_batch = np.array(pad_sentence_batch(sources_batch, source_pad_int))
        pad_targets_batch = np.array(pad_sentence_batch(targets_batch, target_pad_int))

        # Need the lengths for the _lengths parameters
        pad_targets_lengths = []
        for target in pad_targets_batch:
            pad_targets_lengths.append(len(target))

        pad_source_lengths = []
        for source in pad_sources_batch:
            pad_source_lengths.append(len(source))

        yield pad_sources_batch, pad_targets_batch, pad_source_lengths, pad_targets_lengths


### Train
Train the neural network on the preprocessed data. If you have a hard time getting a good loss, check the forms to see if anyone is having the same problem.

In [28]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
def get_accuracy(target, logits):
    """
    Calculate accuracy
    """
    max_seq = max(target.shape[1], logits.shape[1])
    if max_seq - target.shape[1]:
        target = np.pad(
            target,
            [(0,0),(0,max_seq - target.shape[1])],
            'constant')
    if max_seq - logits.shape[1]:
        logits = np.pad(
            logits,
            [(0,0),(0,max_seq - logits.shape[1])],
            'constant')

    return np.mean(np.equal(target, logits))

# Split data to training and validation sets
train_source = source_int_text[batch_size:]
train_target = target_int_text[batch_size:]
valid_source = source_int_text[:batch_size]
valid_target = target_int_text[:batch_size]
(valid_sources_batch, valid_targets_batch, valid_sources_lengths, valid_targets_lengths ) = next(get_batches(valid_source,
                                                                                                             valid_target,
                                                                                                             batch_size,
                                                                                                             source_vocab_to_int['<PAD>'],
                                                                                                             target_vocab_to_int['<PAD>']))                                                                                                  
with tf.Session(graph=train_graph) as sess:
    sess.run(tf.global_variables_initializer())

    for epoch_i in range(epochs):
        for batch_i, (source_batch, target_batch, sources_lengths, targets_lengths) in enumerate(
                get_batches(train_source, train_target, batch_size,
                            source_vocab_to_int['<PAD>'],
                            target_vocab_to_int['<PAD>'])):

            _, loss = sess.run(
                [train_op, cost],
                {input_data: source_batch,
                 targets: target_batch,
                 lr: learning_rate,
                 target_sequence_length: targets_lengths,
                 source_sequence_length: sources_lengths,
                 keep_prob: keep_probability})


            if batch_i % display_step == 0 and batch_i > 0:


                batch_train_logits = sess.run(
                    inference_logits,
                    {input_data: source_batch,
                     source_sequence_length: sources_lengths,
                     target_sequence_length: targets_lengths,
                     keep_prob: 1.0})


                batch_valid_logits = sess.run(
                    inference_logits,
                    {input_data: valid_sources_batch,
                     source_sequence_length: valid_sources_lengths,
                     target_sequence_length: valid_targets_lengths,
                     keep_prob: 1.0})

                train_acc = get_accuracy(target_batch, batch_train_logits)

                valid_acc = get_accuracy(valid_targets_batch, batch_valid_logits)

                print('Epoch {:>3} Batch {:>4}/{} - Train Accuracy: {:>6.4f}, Validation Accuracy: {:>6.4f}, Loss: {:>6.4f}'
                      .format(epoch_i, batch_i, len(source_int_text) // batch_size, train_acc, valid_acc, loss))

    # Save Model
    saver = tf.train.Saver()
    saver.save(sess, save_path)
    print('Model Trained and Saved')

Epoch   0 Batch    1/538 - Train Accuracy: 0.2314, Validation Accuracy: 0.3157, Loss: 5.2305
Epoch   0 Batch    2/538 - Train Accuracy: 0.2523, Validation Accuracy: 0.3171, Loss: 4.7011
Epoch   0 Batch    3/538 - Train Accuracy: 0.2406, Validation Accuracy: 0.3276, Loss: 4.4958
Epoch   0 Batch    4/538 - Train Accuracy: 0.2717, Validation Accuracy: 0.3453, Loss: 4.3308
Epoch   0 Batch    5/538 - Train Accuracy: 0.2978, Validation Accuracy: 0.3464, Loss: 4.0261
Epoch   0 Batch    6/538 - Train Accuracy: 0.3056, Validation Accuracy: 0.3484, Loss: 3.8479
Epoch   0 Batch    7/538 - Train Accuracy: 0.2844, Validation Accuracy: 0.3487, Loss: 3.8574
Epoch   0 Batch    8/538 - Train Accuracy: 0.2803, Validation Accuracy: 0.3473, Loss: 3.7164
Epoch   0 Batch    9/538 - Train Accuracy: 0.2867, Validation Accuracy: 0.3542, Loss: 3.5664
Epoch   0 Batch   10/538 - Train Accuracy: 0.2932, Validation Accuracy: 0.3755, Loss: 3.5813
Epoch   0 Batch   11/538 - Train Accuracy: 0.3232, Validation Accuracy

Epoch   0 Batch   90/538 - Train Accuracy: 0.4667, Validation Accuracy: 0.4975, Loss: 1.6670
Epoch   0 Batch   91/538 - Train Accuracy: 0.4537, Validation Accuracy: 0.5078, Loss: 1.7368
Epoch   0 Batch   92/538 - Train Accuracy: 0.4791, Validation Accuracy: 0.5263, Loss: 1.6939
Epoch   0 Batch   93/538 - Train Accuracy: 0.4322, Validation Accuracy: 0.4961, Loss: 1.7028
Epoch   0 Batch   94/538 - Train Accuracy: 0.4557, Validation Accuracy: 0.5032, Loss: 1.6905
Epoch   0 Batch   95/538 - Train Accuracy: 0.5176, Validation Accuracy: 0.5162, Loss: 1.5392
Epoch   0 Batch   96/538 - Train Accuracy: 0.5121, Validation Accuracy: 0.5238, Loss: 1.5545
Epoch   0 Batch   97/538 - Train Accuracy: 0.4260, Validation Accuracy: 0.4931, Loss: 1.6535
Epoch   0 Batch   98/538 - Train Accuracy: 0.4650, Validation Accuracy: 0.4933, Loss: 1.5379
Epoch   0 Batch   99/538 - Train Accuracy: 0.4807, Validation Accuracy: 0.5293, Loss: 1.6598
Epoch   0 Batch  100/538 - Train Accuracy: 0.4883, Validation Accuracy

Epoch   0 Batch  179/538 - Train Accuracy: 0.5094, Validation Accuracy: 0.5396, Loss: 1.1226
Epoch   0 Batch  180/538 - Train Accuracy: 0.5294, Validation Accuracy: 0.5314, Loss: 1.1016
Epoch   0 Batch  181/538 - Train Accuracy: 0.4920, Validation Accuracy: 0.5552, Loss: 1.1586
Epoch   0 Batch  182/538 - Train Accuracy: 0.5254, Validation Accuracy: 0.5577, Loss: 1.1339
Epoch   0 Batch  183/538 - Train Accuracy: 0.5495, Validation Accuracy: 0.5494, Loss: 1.0508
Epoch   0 Batch  184/538 - Train Accuracy: 0.4760, Validation Accuracy: 0.5123, Loss: 1.0526
Epoch   0 Batch  185/538 - Train Accuracy: 0.5107, Validation Accuracy: 0.5451, Loss: 1.1030
Epoch   0 Batch  186/538 - Train Accuracy: 0.5312, Validation Accuracy: 0.5463, Loss: 1.0695
Epoch   0 Batch  187/538 - Train Accuracy: 0.5662, Validation Accuracy: 0.5543, Loss: 1.0259
Epoch   0 Batch  188/538 - Train Accuracy: 0.4838, Validation Accuracy: 0.5321, Loss: 1.0865
Epoch   0 Batch  189/538 - Train Accuracy: 0.5080, Validation Accuracy

Epoch   0 Batch  268/538 - Train Accuracy: 0.5945, Validation Accuracy: 0.5856, Loss: 0.7649
Epoch   0 Batch  269/538 - Train Accuracy: 0.5613, Validation Accuracy: 0.5879, Loss: 0.8223
Epoch   0 Batch  270/538 - Train Accuracy: 0.5744, Validation Accuracy: 0.5870, Loss: 0.8128
Epoch   0 Batch  271/538 - Train Accuracy: 0.5639, Validation Accuracy: 0.5927, Loss: 0.8190
Epoch   0 Batch  272/538 - Train Accuracy: 0.5586, Validation Accuracy: 0.5925, Loss: 0.8594
Epoch   0 Batch  273/538 - Train Accuracy: 0.5732, Validation Accuracy: 0.5925, Loss: 0.8225
Epoch   0 Batch  274/538 - Train Accuracy: 0.5352, Validation Accuracy: 0.5824, Loss: 0.8516
Epoch   0 Batch  275/538 - Train Accuracy: 0.5582, Validation Accuracy: 0.5820, Loss: 0.8354
Epoch   0 Batch  276/538 - Train Accuracy: 0.5844, Validation Accuracy: 0.5824, Loss: 0.8139
Epoch   0 Batch  277/538 - Train Accuracy: 0.5818, Validation Accuracy: 0.5890, Loss: 0.8039
Epoch   0 Batch  278/538 - Train Accuracy: 0.5627, Validation Accuracy

Epoch   0 Batch  357/538 - Train Accuracy: 0.6080, Validation Accuracy: 0.6179, Loss: 0.6833
Epoch   0 Batch  358/538 - Train Accuracy: 0.5924, Validation Accuracy: 0.6289, Loss: 0.6863
Epoch   0 Batch  359/538 - Train Accuracy: 0.6144, Validation Accuracy: 0.6161, Loss: 0.6747
Epoch   0 Batch  360/538 - Train Accuracy: 0.6025, Validation Accuracy: 0.6213, Loss: 0.6960
Epoch   0 Batch  361/538 - Train Accuracy: 0.6088, Validation Accuracy: 0.6207, Loss: 0.6666
Epoch   0 Batch  362/538 - Train Accuracy: 0.6075, Validation Accuracy: 0.6206, Loss: 0.6552
Epoch   0 Batch  363/538 - Train Accuracy: 0.6124, Validation Accuracy: 0.6135, Loss: 0.6553
Epoch   0 Batch  364/538 - Train Accuracy: 0.5826, Validation Accuracy: 0.6175, Loss: 0.7172
Epoch   0 Batch  365/538 - Train Accuracy: 0.6138, Validation Accuracy: 0.6188, Loss: 0.6856
Epoch   0 Batch  366/538 - Train Accuracy: 0.6066, Validation Accuracy: 0.6230, Loss: 0.6947
Epoch   0 Batch  367/538 - Train Accuracy: 0.6182, Validation Accuracy

Epoch   0 Batch  446/538 - Train Accuracy: 0.6304, Validation Accuracy: 0.6303, Loss: 0.5644
Epoch   0 Batch  447/538 - Train Accuracy: 0.5973, Validation Accuracy: 0.6337, Loss: 0.5967
Epoch   0 Batch  448/538 - Train Accuracy: 0.6337, Validation Accuracy: 0.6369, Loss: 0.5628
Epoch   0 Batch  449/538 - Train Accuracy: 0.6023, Validation Accuracy: 0.6339, Loss: 0.6243
Epoch   0 Batch  450/538 - Train Accuracy: 0.6447, Validation Accuracy: 0.6275, Loss: 0.6029
Epoch   0 Batch  451/538 - Train Accuracy: 0.6121, Validation Accuracy: 0.6239, Loss: 0.5972
Epoch   0 Batch  452/538 - Train Accuracy: 0.6369, Validation Accuracy: 0.6406, Loss: 0.5805
Epoch   0 Batch  453/538 - Train Accuracy: 0.6092, Validation Accuracy: 0.6358, Loss: 0.6022
Epoch   0 Batch  454/538 - Train Accuracy: 0.6376, Validation Accuracy: 0.6348, Loss: 0.5770
Epoch   0 Batch  455/538 - Train Accuracy: 0.6463, Validation Accuracy: 0.6211, Loss: 0.5331
Epoch   0 Batch  456/538 - Train Accuracy: 0.6781, Validation Accuracy

Epoch   0 Batch  535/538 - Train Accuracy: 0.6555, Validation Accuracy: 0.6438, Loss: 0.5143
Epoch   0 Batch  536/538 - Train Accuracy: 0.6375, Validation Accuracy: 0.6420, Loss: 0.5308
Epoch   1 Batch    1/538 - Train Accuracy: 0.6256, Validation Accuracy: 0.6415, Loss: 0.5236
Epoch   1 Batch    2/538 - Train Accuracy: 0.6250, Validation Accuracy: 0.6531, Loss: 0.5483
Epoch   1 Batch    3/538 - Train Accuracy: 0.6209, Validation Accuracy: 0.6222, Loss: 0.5264
Epoch   1 Batch    4/538 - Train Accuracy: 0.6402, Validation Accuracy: 0.6271, Loss: 0.5261
Epoch   1 Batch    5/538 - Train Accuracy: 0.6289, Validation Accuracy: 0.6465, Loss: 0.5248
Epoch   1 Batch    6/538 - Train Accuracy: 0.6559, Validation Accuracy: 0.6413, Loss: 0.5002
Epoch   1 Batch    7/538 - Train Accuracy: 0.6564, Validation Accuracy: 0.6321, Loss: 0.5204
Epoch   1 Batch    8/538 - Train Accuracy: 0.6576, Validation Accuracy: 0.6511, Loss: 0.5246
Epoch   1 Batch    9/538 - Train Accuracy: 0.6555, Validation Accuracy

Epoch   1 Batch   88/538 - Train Accuracy: 0.6975, Validation Accuracy: 0.6857, Loss: 0.4504
Epoch   1 Batch   89/538 - Train Accuracy: 0.6893, Validation Accuracy: 0.6777, Loss: 0.4410
Epoch   1 Batch   90/538 - Train Accuracy: 0.6789, Validation Accuracy: 0.6742, Loss: 0.4446
Epoch   1 Batch   91/538 - Train Accuracy: 0.6797, Validation Accuracy: 0.6737, Loss: 0.4490
Epoch   1 Batch   92/538 - Train Accuracy: 0.6740, Validation Accuracy: 0.6768, Loss: 0.4529
Epoch   1 Batch   93/538 - Train Accuracy: 0.6699, Validation Accuracy: 0.6838, Loss: 0.4591
Epoch   1 Batch   94/538 - Train Accuracy: 0.6977, Validation Accuracy: 0.6772, Loss: 0.4402
Epoch   1 Batch   95/538 - Train Accuracy: 0.7100, Validation Accuracy: 0.6788, Loss: 0.4032
Epoch   1 Batch   96/538 - Train Accuracy: 0.7137, Validation Accuracy: 0.6866, Loss: 0.4108
Epoch   1 Batch   97/538 - Train Accuracy: 0.6805, Validation Accuracy: 0.6838, Loss: 0.4367
Epoch   1 Batch   98/538 - Train Accuracy: 0.7217, Validation Accuracy

Epoch   1 Batch  177/538 - Train Accuracy: 0.7422, Validation Accuracy: 0.7275, Loss: 0.3590
Epoch   1 Batch  178/538 - Train Accuracy: 0.7210, Validation Accuracy: 0.7290, Loss: 0.3454
Epoch   1 Batch  179/538 - Train Accuracy: 0.7230, Validation Accuracy: 0.7287, Loss: 0.3655
Epoch   1 Batch  180/538 - Train Accuracy: 0.7452, Validation Accuracy: 0.7324, Loss: 0.3551
Epoch   1 Batch  181/538 - Train Accuracy: 0.7021, Validation Accuracy: 0.7141, Loss: 0.3778
Epoch   1 Batch  182/538 - Train Accuracy: 0.7109, Validation Accuracy: 0.7244, Loss: 0.3722
Epoch   1 Batch  183/538 - Train Accuracy: 0.7561, Validation Accuracy: 0.7228, Loss: 0.3382
Epoch   1 Batch  184/538 - Train Accuracy: 0.7439, Validation Accuracy: 0.7182, Loss: 0.3464
Epoch   1 Batch  185/538 - Train Accuracy: 0.7461, Validation Accuracy: 0.7324, Loss: 0.3496
Epoch   1 Batch  186/538 - Train Accuracy: 0.7208, Validation Accuracy: 0.7340, Loss: 0.3498
Epoch   1 Batch  187/538 - Train Accuracy: 0.7731, Validation Accuracy

Epoch   1 Batch  266/538 - Train Accuracy: 0.7922, Validation Accuracy: 0.7701, Loss: 0.2844
Epoch   1 Batch  267/538 - Train Accuracy: 0.8088, Validation Accuracy: 0.7775, Loss: 0.2859
Epoch   1 Batch  268/538 - Train Accuracy: 0.8181, Validation Accuracy: 0.7837, Loss: 0.2594
Epoch   1 Batch  269/538 - Train Accuracy: 0.8051, Validation Accuracy: 0.7786, Loss: 0.2752
Epoch   1 Batch  270/538 - Train Accuracy: 0.7797, Validation Accuracy: 0.7750, Loss: 0.2864
Epoch   1 Batch  271/538 - Train Accuracy: 0.7937, Validation Accuracy: 0.7795, Loss: 0.2738
Epoch   1 Batch  272/538 - Train Accuracy: 0.7666, Validation Accuracy: 0.7876, Loss: 0.2973
Epoch   1 Batch  273/538 - Train Accuracy: 0.8059, Validation Accuracy: 0.7999, Loss: 0.2801
Epoch   1 Batch  274/538 - Train Accuracy: 0.7441, Validation Accuracy: 0.8082, Loss: 0.2934
Epoch   1 Batch  275/538 - Train Accuracy: 0.7930, Validation Accuracy: 0.7917, Loss: 0.2898
Epoch   1 Batch  276/538 - Train Accuracy: 0.8189, Validation Accuracy

Epoch   1 Batch  355/538 - Train Accuracy: 0.8559, Validation Accuracy: 0.8386, Loss: 0.2234
Epoch   1 Batch  356/538 - Train Accuracy: 0.8406, Validation Accuracy: 0.8221, Loss: 0.1903
Epoch   1 Batch  357/538 - Train Accuracy: 0.8602, Validation Accuracy: 0.8145, Loss: 0.2158
Epoch   1 Batch  358/538 - Train Accuracy: 0.8223, Validation Accuracy: 0.8374, Loss: 0.2096
Epoch   1 Batch  359/538 - Train Accuracy: 0.8354, Validation Accuracy: 0.8420, Loss: 0.2088
Epoch   1 Batch  360/538 - Train Accuracy: 0.8529, Validation Accuracy: 0.8439, Loss: 0.2179
Epoch   1 Batch  361/538 - Train Accuracy: 0.8363, Validation Accuracy: 0.8224, Loss: 0.2163
Epoch   1 Batch  362/538 - Train Accuracy: 0.8439, Validation Accuracy: 0.8171, Loss: 0.1937
Epoch   1 Batch  363/538 - Train Accuracy: 0.8270, Validation Accuracy: 0.8260, Loss: 0.2074
Epoch   1 Batch  364/538 - Train Accuracy: 0.8287, Validation Accuracy: 0.8386, Loss: 0.2275
Epoch   1 Batch  365/538 - Train Accuracy: 0.8278, Validation Accuracy

Epoch   1 Batch  444/538 - Train Accuracy: 0.8942, Validation Accuracy: 0.8704, Loss: 0.1520
Epoch   1 Batch  445/538 - Train Accuracy: 0.8949, Validation Accuracy: 0.8695, Loss: 0.1550
Epoch   1 Batch  446/538 - Train Accuracy: 0.8715, Validation Accuracy: 0.8675, Loss: 0.1505
Epoch   1 Batch  447/538 - Train Accuracy: 0.8828, Validation Accuracy: 0.8597, Loss: 0.1601
Epoch   1 Batch  448/538 - Train Accuracy: 0.8849, Validation Accuracy: 0.8491, Loss: 0.1436
Epoch   1 Batch  449/538 - Train Accuracy: 0.8852, Validation Accuracy: 0.8482, Loss: 0.1674
Epoch   1 Batch  450/538 - Train Accuracy: 0.8579, Validation Accuracy: 0.8631, Loss: 0.1700
Epoch   1 Batch  451/538 - Train Accuracy: 0.8592, Validation Accuracy: 0.8555, Loss: 0.1591
Epoch   1 Batch  452/538 - Train Accuracy: 0.8943, Validation Accuracy: 0.8532, Loss: 0.1462
Epoch   1 Batch  453/538 - Train Accuracy: 0.8801, Validation Accuracy: 0.8528, Loss: 0.1683
Epoch   1 Batch  454/538 - Train Accuracy: 0.8802, Validation Accuracy

Epoch   1 Batch  533/538 - Train Accuracy: 0.9119, Validation Accuracy: 0.8915, Loss: 0.1256
Epoch   1 Batch  534/538 - Train Accuracy: 0.9035, Validation Accuracy: 0.8826, Loss: 0.1131
Epoch   1 Batch  535/538 - Train Accuracy: 0.8986, Validation Accuracy: 0.8723, Loss: 0.1164
Epoch   1 Batch  536/538 - Train Accuracy: 0.8925, Validation Accuracy: 0.8729, Loss: 0.1406
Epoch   2 Batch    1/538 - Train Accuracy: 0.9211, Validation Accuracy: 0.8761, Loss: 0.1229
Epoch   2 Batch    2/538 - Train Accuracy: 0.8789, Validation Accuracy: 0.8722, Loss: 0.1407
Epoch   2 Batch    3/538 - Train Accuracy: 0.9037, Validation Accuracy: 0.8707, Loss: 0.1179
Epoch   2 Batch    4/538 - Train Accuracy: 0.9047, Validation Accuracy: 0.8805, Loss: 0.1168
Epoch   2 Batch    5/538 - Train Accuracy: 0.8897, Validation Accuracy: 0.8899, Loss: 0.1225
Epoch   2 Batch    6/538 - Train Accuracy: 0.9029, Validation Accuracy: 0.8899, Loss: 0.1157
Epoch   2 Batch    7/538 - Train Accuracy: 0.9287, Validation Accuracy

Epoch   2 Batch   86/538 - Train Accuracy: 0.9068, Validation Accuracy: 0.8862, Loss: 0.0969
Epoch   2 Batch   87/538 - Train Accuracy: 0.9160, Validation Accuracy: 0.8924, Loss: 0.0993
Epoch   2 Batch   88/538 - Train Accuracy: 0.9000, Validation Accuracy: 0.8960, Loss: 0.0997
Epoch   2 Batch   89/538 - Train Accuracy: 0.9125, Validation Accuracy: 0.8857, Loss: 0.0948
Epoch   2 Batch   90/538 - Train Accuracy: 0.8919, Validation Accuracy: 0.8754, Loss: 0.1045
Epoch   2 Batch   91/538 - Train Accuracy: 0.9041, Validation Accuracy: 0.8885, Loss: 0.0950
Epoch   2 Batch   92/538 - Train Accuracy: 0.8959, Validation Accuracy: 0.8929, Loss: 0.0980
Epoch   2 Batch   93/538 - Train Accuracy: 0.8834, Validation Accuracy: 0.8983, Loss: 0.0939
Epoch   2 Batch   94/538 - Train Accuracy: 0.9254, Validation Accuracy: 0.8940, Loss: 0.0929
Epoch   2 Batch   95/538 - Train Accuracy: 0.8944, Validation Accuracy: 0.8860, Loss: 0.0884
Epoch   2 Batch   96/538 - Train Accuracy: 0.9340, Validation Accuracy

Epoch   2 Batch  175/538 - Train Accuracy: 0.9008, Validation Accuracy: 0.8764, Loss: 0.0823
Epoch   2 Batch  176/538 - Train Accuracy: 0.9203, Validation Accuracy: 0.8826, Loss: 0.0974
Epoch   2 Batch  177/538 - Train Accuracy: 0.9081, Validation Accuracy: 0.8951, Loss: 0.0891
Epoch   2 Batch  178/538 - Train Accuracy: 0.8917, Validation Accuracy: 0.8908, Loss: 0.0852
Epoch   2 Batch  179/538 - Train Accuracy: 0.9187, Validation Accuracy: 0.8938, Loss: 0.0840
Epoch   2 Batch  180/538 - Train Accuracy: 0.9176, Validation Accuracy: 0.8846, Loss: 0.0825
Epoch   2 Batch  181/538 - Train Accuracy: 0.8918, Validation Accuracy: 0.8912, Loss: 0.0979
Epoch   2 Batch  182/538 - Train Accuracy: 0.9209, Validation Accuracy: 0.8848, Loss: 0.0781
Epoch   2 Batch  183/538 - Train Accuracy: 0.9269, Validation Accuracy: 0.8917, Loss: 0.0739
Epoch   2 Batch  184/538 - Train Accuracy: 0.9157, Validation Accuracy: 0.8906, Loss: 0.0845
Epoch   2 Batch  185/538 - Train Accuracy: 0.9359, Validation Accuracy

Epoch   2 Batch  264/538 - Train Accuracy: 0.9035, Validation Accuracy: 0.9098, Loss: 0.0721
Epoch   2 Batch  265/538 - Train Accuracy: 0.9121, Validation Accuracy: 0.9002, Loss: 0.0793
Epoch   2 Batch  266/538 - Train Accuracy: 0.9049, Validation Accuracy: 0.9018, Loss: 0.0716
Epoch   2 Batch  267/538 - Train Accuracy: 0.9084, Validation Accuracy: 0.9102, Loss: 0.0759
Epoch   2 Batch  268/538 - Train Accuracy: 0.9427, Validation Accuracy: 0.9178, Loss: 0.0574
Epoch   2 Batch  269/538 - Train Accuracy: 0.9215, Validation Accuracy: 0.9134, Loss: 0.0752
Epoch   2 Batch  270/538 - Train Accuracy: 0.9141, Validation Accuracy: 0.9176, Loss: 0.0662
Epoch   2 Batch  271/538 - Train Accuracy: 0.9320, Validation Accuracy: 0.9229, Loss: 0.0597
Epoch   2 Batch  272/538 - Train Accuracy: 0.9277, Validation Accuracy: 0.9178, Loss: 0.0793
Epoch   2 Batch  273/538 - Train Accuracy: 0.9264, Validation Accuracy: 0.9221, Loss: 0.0717
Epoch   2 Batch  274/538 - Train Accuracy: 0.8869, Validation Accuracy

Epoch   2 Batch  353/538 - Train Accuracy: 0.9146, Validation Accuracy: 0.9254, Loss: 0.0693
Epoch   2 Batch  354/538 - Train Accuracy: 0.9057, Validation Accuracy: 0.9270, Loss: 0.0686
Epoch   2 Batch  355/538 - Train Accuracy: 0.9311, Validation Accuracy: 0.9304, Loss: 0.0640
Epoch   2 Batch  356/538 - Train Accuracy: 0.9274, Validation Accuracy: 0.9274, Loss: 0.0544
Epoch   2 Batch  357/538 - Train Accuracy: 0.9408, Validation Accuracy: 0.9279, Loss: 0.0616
Epoch   2 Batch  358/538 - Train Accuracy: 0.9389, Validation Accuracy: 0.9318, Loss: 0.0571
Epoch   2 Batch  359/538 - Train Accuracy: 0.9232, Validation Accuracy: 0.9265, Loss: 0.0625
Epoch   2 Batch  360/538 - Train Accuracy: 0.9219, Validation Accuracy: 0.9334, Loss: 0.0598
Epoch   2 Batch  361/538 - Train Accuracy: 0.9208, Validation Accuracy: 0.9267, Loss: 0.0626
Epoch   2 Batch  362/538 - Train Accuracy: 0.9358, Validation Accuracy: 0.9270, Loss: 0.0540
Epoch   2 Batch  363/538 - Train Accuracy: 0.9221, Validation Accuracy

Epoch   2 Batch  442/538 - Train Accuracy: 0.9354, Validation Accuracy: 0.9304, Loss: 0.0443
Epoch   2 Batch  443/538 - Train Accuracy: 0.9312, Validation Accuracy: 0.9237, Loss: 0.0569
Epoch   2 Batch  444/538 - Train Accuracy: 0.9373, Validation Accuracy: 0.9286, Loss: 0.0520
Epoch   2 Batch  445/538 - Train Accuracy: 0.9551, Validation Accuracy: 0.9332, Loss: 0.0483
Epoch   2 Batch  446/538 - Train Accuracy: 0.9414, Validation Accuracy: 0.9355, Loss: 0.0525
Epoch   2 Batch  447/538 - Train Accuracy: 0.9322, Validation Accuracy: 0.9306, Loss: 0.0553
Epoch   2 Batch  448/538 - Train Accuracy: 0.9353, Validation Accuracy: 0.9300, Loss: 0.0492
Epoch   2 Batch  449/538 - Train Accuracy: 0.9457, Validation Accuracy: 0.9284, Loss: 0.0603
Epoch   2 Batch  450/538 - Train Accuracy: 0.9172, Validation Accuracy: 0.9283, Loss: 0.0668
Epoch   2 Batch  451/538 - Train Accuracy: 0.9256, Validation Accuracy: 0.9336, Loss: 0.0540
Epoch   2 Batch  452/538 - Train Accuracy: 0.9447, Validation Accuracy

Epoch   2 Batch  531/538 - Train Accuracy: 0.9256, Validation Accuracy: 0.9460, Loss: 0.0547
Epoch   2 Batch  532/538 - Train Accuracy: 0.9330, Validation Accuracy: 0.9426, Loss: 0.0452
Epoch   2 Batch  533/538 - Train Accuracy: 0.9484, Validation Accuracy: 0.9379, Loss: 0.0458
Epoch   2 Batch  534/538 - Train Accuracy: 0.9448, Validation Accuracy: 0.9364, Loss: 0.0399
Epoch   2 Batch  535/538 - Train Accuracy: 0.9395, Validation Accuracy: 0.9391, Loss: 0.0463
Epoch   2 Batch  536/538 - Train Accuracy: 0.9427, Validation Accuracy: 0.9341, Loss: 0.0603
Epoch   3 Batch    1/538 - Train Accuracy: 0.9469, Validation Accuracy: 0.9276, Loss: 0.0498
Epoch   3 Batch    2/538 - Train Accuracy: 0.9205, Validation Accuracy: 0.9332, Loss: 0.0627
Epoch   3 Batch    3/538 - Train Accuracy: 0.9525, Validation Accuracy: 0.9297, Loss: 0.0491
Epoch   3 Batch    4/538 - Train Accuracy: 0.9406, Validation Accuracy: 0.9324, Loss: 0.0484
Epoch   3 Batch    5/538 - Train Accuracy: 0.9327, Validation Accuracy

Epoch   3 Batch   84/538 - Train Accuracy: 0.9498, Validation Accuracy: 0.9359, Loss: 0.0469
Epoch   3 Batch   85/538 - Train Accuracy: 0.9579, Validation Accuracy: 0.9345, Loss: 0.0389
Epoch   3 Batch   86/538 - Train Accuracy: 0.9488, Validation Accuracy: 0.9325, Loss: 0.0429
Epoch   3 Batch   87/538 - Train Accuracy: 0.9516, Validation Accuracy: 0.9437, Loss: 0.0473
Epoch   3 Batch   88/538 - Train Accuracy: 0.9367, Validation Accuracy: 0.9437, Loss: 0.0447
Epoch   3 Batch   89/538 - Train Accuracy: 0.9467, Validation Accuracy: 0.9426, Loss: 0.0427
Epoch   3 Batch   90/538 - Train Accuracy: 0.9377, Validation Accuracy: 0.9450, Loss: 0.0495
Epoch   3 Batch   91/538 - Train Accuracy: 0.9346, Validation Accuracy: 0.9416, Loss: 0.0443
Epoch   3 Batch   92/538 - Train Accuracy: 0.9422, Validation Accuracy: 0.9347, Loss: 0.0478
Epoch   3 Batch   93/538 - Train Accuracy: 0.9412, Validation Accuracy: 0.9535, Loss: 0.0438
Epoch   3 Batch   94/538 - Train Accuracy: 0.9498, Validation Accuracy

Epoch   3 Batch  173/538 - Train Accuracy: 0.9632, Validation Accuracy: 0.9347, Loss: 0.0377
Epoch   3 Batch  174/538 - Train Accuracy: 0.9451, Validation Accuracy: 0.9366, Loss: 0.0422
Epoch   3 Batch  175/538 - Train Accuracy: 0.9492, Validation Accuracy: 0.9448, Loss: 0.0380
Epoch   3 Batch  176/538 - Train Accuracy: 0.9529, Validation Accuracy: 0.9508, Loss: 0.0490
Epoch   3 Batch  177/538 - Train Accuracy: 0.9518, Validation Accuracy: 0.9489, Loss: 0.0460
Epoch   3 Batch  178/538 - Train Accuracy: 0.9196, Validation Accuracy: 0.9489, Loss: 0.0447
Epoch   3 Batch  179/538 - Train Accuracy: 0.9594, Validation Accuracy: 0.9437, Loss: 0.0405
Epoch   3 Batch  180/538 - Train Accuracy: 0.9356, Validation Accuracy: 0.9432, Loss: 0.0452
Epoch   3 Batch  181/538 - Train Accuracy: 0.9279, Validation Accuracy: 0.9425, Loss: 0.0464
Epoch   3 Batch  182/538 - Train Accuracy: 0.9518, Validation Accuracy: 0.9393, Loss: 0.0380
Epoch   3 Batch  183/538 - Train Accuracy: 0.9676, Validation Accuracy

Epoch   3 Batch  262/538 - Train Accuracy: 0.9705, Validation Accuracy: 0.9547, Loss: 0.0372
Epoch   3 Batch  263/538 - Train Accuracy: 0.9453, Validation Accuracy: 0.9595, Loss: 0.0403
Epoch   3 Batch  264/538 - Train Accuracy: 0.9328, Validation Accuracy: 0.9558, Loss: 0.0407
Epoch   3 Batch  265/538 - Train Accuracy: 0.9406, Validation Accuracy: 0.9490, Loss: 0.0495
Epoch   3 Batch  266/538 - Train Accuracy: 0.9314, Validation Accuracy: 0.9425, Loss: 0.0407
Epoch   3 Batch  267/538 - Train Accuracy: 0.9193, Validation Accuracy: 0.9379, Loss: 0.0409
Epoch   3 Batch  268/538 - Train Accuracy: 0.9756, Validation Accuracy: 0.9474, Loss: 0.0289
Epoch   3 Batch  269/538 - Train Accuracy: 0.9559, Validation Accuracy: 0.9556, Loss: 0.0435
Epoch   3 Batch  270/538 - Train Accuracy: 0.9484, Validation Accuracy: 0.9529, Loss: 0.0329
Epoch   3 Batch  271/538 - Train Accuracy: 0.9563, Validation Accuracy: 0.9490, Loss: 0.0323
Epoch   3 Batch  272/538 - Train Accuracy: 0.9324, Validation Accuracy

Epoch   3 Batch  351/538 - Train Accuracy: 0.9396, Validation Accuracy: 0.9558, Loss: 0.0438
Epoch   3 Batch  352/538 - Train Accuracy: 0.9435, Validation Accuracy: 0.9528, Loss: 0.0508
Epoch   3 Batch  353/538 - Train Accuracy: 0.9314, Validation Accuracy: 0.9483, Loss: 0.0405
Epoch   3 Batch  354/538 - Train Accuracy: 0.9377, Validation Accuracy: 0.9478, Loss: 0.0413
Epoch   3 Batch  355/538 - Train Accuracy: 0.9570, Validation Accuracy: 0.9455, Loss: 0.0382
Epoch   3 Batch  356/538 - Train Accuracy: 0.9581, Validation Accuracy: 0.9483, Loss: 0.0320
Epoch   3 Batch  357/538 - Train Accuracy: 0.9588, Validation Accuracy: 0.9569, Loss: 0.0369
Epoch   3 Batch  358/538 - Train Accuracy: 0.9639, Validation Accuracy: 0.9533, Loss: 0.0326
Epoch   3 Batch  359/538 - Train Accuracy: 0.9496, Validation Accuracy: 0.9535, Loss: 0.0364
Epoch   3 Batch  360/538 - Train Accuracy: 0.9496, Validation Accuracy: 0.9526, Loss: 0.0348
Epoch   3 Batch  361/538 - Train Accuracy: 0.9529, Validation Accuracy

Epoch   3 Batch  440/538 - Train Accuracy: 0.9641, Validation Accuracy: 0.9590, Loss: 0.0357
Epoch   3 Batch  441/538 - Train Accuracy: 0.9365, Validation Accuracy: 0.9576, Loss: 0.0401
Epoch   3 Batch  442/538 - Train Accuracy: 0.9553, Validation Accuracy: 0.9499, Loss: 0.0272
Epoch   3 Batch  443/538 - Train Accuracy: 0.9405, Validation Accuracy: 0.9384, Loss: 0.0340
Epoch   3 Batch  444/538 - Train Accuracy: 0.9574, Validation Accuracy: 0.9350, Loss: 0.0328
Epoch   3 Batch  445/538 - Train Accuracy: 0.9662, Validation Accuracy: 0.9499, Loss: 0.0289
Epoch   3 Batch  446/538 - Train Accuracy: 0.9600, Validation Accuracy: 0.9499, Loss: 0.0293
Epoch   3 Batch  447/538 - Train Accuracy: 0.9531, Validation Accuracy: 0.9503, Loss: 0.0307
Epoch   3 Batch  448/538 - Train Accuracy: 0.9494, Validation Accuracy: 0.9439, Loss: 0.0284
Epoch   3 Batch  449/538 - Train Accuracy: 0.9646, Validation Accuracy: 0.9565, Loss: 0.0381
Epoch   3 Batch  450/538 - Train Accuracy: 0.9440, Validation Accuracy

Epoch   3 Batch  529/538 - Train Accuracy: 0.9400, Validation Accuracy: 0.9519, Loss: 0.0306
Epoch   3 Batch  530/538 - Train Accuracy: 0.9537, Validation Accuracy: 0.9593, Loss: 0.0396
Epoch   3 Batch  531/538 - Train Accuracy: 0.9576, Validation Accuracy: 0.9577, Loss: 0.0357
Epoch   3 Batch  532/538 - Train Accuracy: 0.9600, Validation Accuracy: 0.9579, Loss: 0.0261
Epoch   3 Batch  533/538 - Train Accuracy: 0.9645, Validation Accuracy: 0.9574, Loss: 0.0305
Epoch   3 Batch  534/538 - Train Accuracy: 0.9674, Validation Accuracy: 0.9586, Loss: 0.0245
Epoch   3 Batch  535/538 - Train Accuracy: 0.9572, Validation Accuracy: 0.9590, Loss: 0.0299
Epoch   3 Batch  536/538 - Train Accuracy: 0.9777, Validation Accuracy: 0.9513, Loss: 0.0357
Epoch   4 Batch    1/538 - Train Accuracy: 0.9678, Validation Accuracy: 0.9558, Loss: 0.0323
Epoch   4 Batch    2/538 - Train Accuracy: 0.9648, Validation Accuracy: 0.9460, Loss: 0.0392
Epoch   4 Batch    3/538 - Train Accuracy: 0.9701, Validation Accuracy

Epoch   4 Batch   82/538 - Train Accuracy: 0.9645, Validation Accuracy: 0.9654, Loss: 0.0290
Epoch   4 Batch   83/538 - Train Accuracy: 0.9471, Validation Accuracy: 0.9650, Loss: 0.0300
Epoch   4 Batch   84/538 - Train Accuracy: 0.9667, Validation Accuracy: 0.9627, Loss: 0.0300
Epoch   4 Batch   85/538 - Train Accuracy: 0.9803, Validation Accuracy: 0.9602, Loss: 0.0242
Epoch   4 Batch   86/538 - Train Accuracy: 0.9686, Validation Accuracy: 0.9553, Loss: 0.0276
Epoch   4 Batch   87/538 - Train Accuracy: 0.9600, Validation Accuracy: 0.9602, Loss: 0.0308
Epoch   4 Batch   88/538 - Train Accuracy: 0.9668, Validation Accuracy: 0.9529, Loss: 0.0296
Epoch   4 Batch   89/538 - Train Accuracy: 0.9564, Validation Accuracy: 0.9489, Loss: 0.0274
Epoch   4 Batch   90/538 - Train Accuracy: 0.9602, Validation Accuracy: 0.9561, Loss: 0.0336
Epoch   4 Batch   91/538 - Train Accuracy: 0.9604, Validation Accuracy: 0.9650, Loss: 0.0316
Epoch   4 Batch   92/538 - Train Accuracy: 0.9578, Validation Accuracy

Epoch   4 Batch  171/538 - Train Accuracy: 0.9586, Validation Accuracy: 0.9538, Loss: 0.0334
Epoch   4 Batch  172/538 - Train Accuracy: 0.9674, Validation Accuracy: 0.9606, Loss: 0.0258
Epoch   4 Batch  173/538 - Train Accuracy: 0.9689, Validation Accuracy: 0.9533, Loss: 0.0248
Epoch   4 Batch  174/538 - Train Accuracy: 0.9633, Validation Accuracy: 0.9428, Loss: 0.0276
Epoch   4 Batch  175/538 - Train Accuracy: 0.9660, Validation Accuracy: 0.9444, Loss: 0.0255
Epoch   4 Batch  176/538 - Train Accuracy: 0.9682, Validation Accuracy: 0.9517, Loss: 0.0338
Epoch   4 Batch  177/538 - Train Accuracy: 0.9695, Validation Accuracy: 0.9542, Loss: 0.0287
Epoch   4 Batch  178/538 - Train Accuracy: 0.9457, Validation Accuracy: 0.9556, Loss: 0.0289
Epoch   4 Batch  179/538 - Train Accuracy: 0.9656, Validation Accuracy: 0.9577, Loss: 0.0267
Epoch   4 Batch  180/538 - Train Accuracy: 0.9529, Validation Accuracy: 0.9558, Loss: 0.0296
Epoch   4 Batch  181/538 - Train Accuracy: 0.9506, Validation Accuracy

Epoch   4 Batch  260/538 - Train Accuracy: 0.9459, Validation Accuracy: 0.9609, Loss: 0.0288
Epoch   4 Batch  261/538 - Train Accuracy: 0.9736, Validation Accuracy: 0.9629, Loss: 0.0298
Epoch   4 Batch  262/538 - Train Accuracy: 0.9734, Validation Accuracy: 0.9620, Loss: 0.0250
Epoch   4 Batch  263/538 - Train Accuracy: 0.9635, Validation Accuracy: 0.9561, Loss: 0.0255
Epoch   4 Batch  264/538 - Train Accuracy: 0.9473, Validation Accuracy: 0.9574, Loss: 0.0279
Epoch   4 Batch  265/538 - Train Accuracy: 0.9531, Validation Accuracy: 0.9570, Loss: 0.0325
Epoch   4 Batch  266/538 - Train Accuracy: 0.9501, Validation Accuracy: 0.9618, Loss: 0.0264
Epoch   4 Batch  267/538 - Train Accuracy: 0.9639, Validation Accuracy: 0.9600, Loss: 0.0275
Epoch   4 Batch  268/538 - Train Accuracy: 0.9753, Validation Accuracy: 0.9595, Loss: 0.0188
Epoch   4 Batch  269/538 - Train Accuracy: 0.9596, Validation Accuracy: 0.9624, Loss: 0.0312
Epoch   4 Batch  270/538 - Train Accuracy: 0.9590, Validation Accuracy

Epoch   4 Batch  349/538 - Train Accuracy: 0.9748, Validation Accuracy: 0.9579, Loss: 0.0203
Epoch   4 Batch  350/538 - Train Accuracy: 0.9686, Validation Accuracy: 0.9538, Loss: 0.0310
Epoch   4 Batch  351/538 - Train Accuracy: 0.9490, Validation Accuracy: 0.9595, Loss: 0.0307
Epoch   4 Batch  352/538 - Train Accuracy: 0.9544, Validation Accuracy: 0.9620, Loss: 0.0379
Epoch   4 Batch  353/538 - Train Accuracy: 0.9592, Validation Accuracy: 0.9659, Loss: 0.0281
Epoch   4 Batch  354/538 - Train Accuracy: 0.9521, Validation Accuracy: 0.9636, Loss: 0.0288
Epoch   4 Batch  355/538 - Train Accuracy: 0.9713, Validation Accuracy: 0.9631, Loss: 0.0256
Epoch   4 Batch  356/538 - Train Accuracy: 0.9714, Validation Accuracy: 0.9606, Loss: 0.0214
Epoch   4 Batch  357/538 - Train Accuracy: 0.9576, Validation Accuracy: 0.9638, Loss: 0.0249
Epoch   4 Batch  358/538 - Train Accuracy: 0.9738, Validation Accuracy: 0.9629, Loss: 0.0215
Epoch   4 Batch  359/538 - Train Accuracy: 0.9561, Validation Accuracy

Epoch   4 Batch  438/538 - Train Accuracy: 0.9711, Validation Accuracy: 0.9634, Loss: 0.0215
Epoch   4 Batch  439/538 - Train Accuracy: 0.9732, Validation Accuracy: 0.9632, Loss: 0.0252
Epoch   4 Batch  440/538 - Train Accuracy: 0.9729, Validation Accuracy: 0.9657, Loss: 0.0261
Epoch   4 Batch  441/538 - Train Accuracy: 0.9500, Validation Accuracy: 0.9600, Loss: 0.0300
Epoch   4 Batch  442/538 - Train Accuracy: 0.9759, Validation Accuracy: 0.9643, Loss: 0.0207
Epoch   4 Batch  443/538 - Train Accuracy: 0.9507, Validation Accuracy: 0.9592, Loss: 0.0252
Epoch   4 Batch  444/538 - Train Accuracy: 0.9647, Validation Accuracy: 0.9577, Loss: 0.0228
Epoch   4 Batch  445/538 - Train Accuracy: 0.9742, Validation Accuracy: 0.9583, Loss: 0.0189
Epoch   4 Batch  446/538 - Train Accuracy: 0.9652, Validation Accuracy: 0.9585, Loss: 0.0216
Epoch   4 Batch  447/538 - Train Accuracy: 0.9582, Validation Accuracy: 0.9558, Loss: 0.0226
Epoch   4 Batch  448/538 - Train Accuracy: 0.9714, Validation Accuracy

Epoch   4 Batch  527/538 - Train Accuracy: 0.9676, Validation Accuracy: 0.9579, Loss: 0.0242
Epoch   4 Batch  528/538 - Train Accuracy: 0.9613, Validation Accuracy: 0.9656, Loss: 0.0260
Epoch   4 Batch  529/538 - Train Accuracy: 0.9521, Validation Accuracy: 0.9670, Loss: 0.0237
Epoch   4 Batch  530/538 - Train Accuracy: 0.9568, Validation Accuracy: 0.9673, Loss: 0.0320
Epoch   4 Batch  531/538 - Train Accuracy: 0.9678, Validation Accuracy: 0.9711, Loss: 0.0261
Epoch   4 Batch  532/538 - Train Accuracy: 0.9730, Validation Accuracy: 0.9672, Loss: 0.0206
Epoch   4 Batch  533/538 - Train Accuracy: 0.9762, Validation Accuracy: 0.9583, Loss: 0.0208
Epoch   4 Batch  534/538 - Train Accuracy: 0.9661, Validation Accuracy: 0.9558, Loss: 0.0182
Epoch   4 Batch  535/538 - Train Accuracy: 0.9635, Validation Accuracy: 0.9595, Loss: 0.0233
Epoch   4 Batch  536/538 - Train Accuracy: 0.9836, Validation Accuracy: 0.9703, Loss: 0.0274
Epoch   5 Batch    1/538 - Train Accuracy: 0.9775, Validation Accuracy

Epoch   5 Batch   80/538 - Train Accuracy: 0.9709, Validation Accuracy: 0.9721, Loss: 0.0207
Epoch   5 Batch   81/538 - Train Accuracy: 0.9711, Validation Accuracy: 0.9661, Loss: 0.0242
Epoch   5 Batch   82/538 - Train Accuracy: 0.9758, Validation Accuracy: 0.9632, Loss: 0.0224
Epoch   5 Batch   83/538 - Train Accuracy: 0.9656, Validation Accuracy: 0.9593, Loss: 0.0239
Epoch   5 Batch   84/538 - Train Accuracy: 0.9734, Validation Accuracy: 0.9570, Loss: 0.0235
Epoch   5 Batch   85/538 - Train Accuracy: 0.9865, Validation Accuracy: 0.9581, Loss: 0.0165
Epoch   5 Batch   86/538 - Train Accuracy: 0.9809, Validation Accuracy: 0.9572, Loss: 0.0191
Epoch   5 Batch   87/538 - Train Accuracy: 0.9609, Validation Accuracy: 0.9599, Loss: 0.0212
Epoch   5 Batch   88/538 - Train Accuracy: 0.9668, Validation Accuracy: 0.9602, Loss: 0.0224
Epoch   5 Batch   89/538 - Train Accuracy: 0.9660, Validation Accuracy: 0.9609, Loss: 0.0207
Epoch   5 Batch   90/538 - Train Accuracy: 0.9688, Validation Accuracy

Epoch   5 Batch  169/538 - Train Accuracy: 0.9805, Validation Accuracy: 0.9659, Loss: 0.0174
Epoch   5 Batch  170/538 - Train Accuracy: 0.9660, Validation Accuracy: 0.9659, Loss: 0.0216
Epoch   5 Batch  171/538 - Train Accuracy: 0.9668, Validation Accuracy: 0.9661, Loss: 0.0236
Epoch   5 Batch  172/538 - Train Accuracy: 0.9788, Validation Accuracy: 0.9604, Loss: 0.0181
Epoch   5 Batch  173/538 - Train Accuracy: 0.9782, Validation Accuracy: 0.9560, Loss: 0.0171
Epoch   5 Batch  174/538 - Train Accuracy: 0.9627, Validation Accuracy: 0.9572, Loss: 0.0201
Epoch   5 Batch  175/538 - Train Accuracy: 0.9738, Validation Accuracy: 0.9576, Loss: 0.0190
Epoch   5 Batch  176/538 - Train Accuracy: 0.9750, Validation Accuracy: 0.9567, Loss: 0.0253
Epoch   5 Batch  177/538 - Train Accuracy: 0.9743, Validation Accuracy: 0.9643, Loss: 0.0239
Epoch   5 Batch  178/538 - Train Accuracy: 0.9580, Validation Accuracy: 0.9652, Loss: 0.0230
Epoch   5 Batch  179/538 - Train Accuracy: 0.9811, Validation Accuracy

Epoch   5 Batch  258/538 - Train Accuracy: 0.9814, Validation Accuracy: 0.9599, Loss: 0.0192
Epoch   5 Batch  259/538 - Train Accuracy: 0.9810, Validation Accuracy: 0.9606, Loss: 0.0168
Epoch   5 Batch  260/538 - Train Accuracy: 0.9652, Validation Accuracy: 0.9561, Loss: 0.0202
Epoch   5 Batch  261/538 - Train Accuracy: 0.9715, Validation Accuracy: 0.9561, Loss: 0.0207
Epoch   5 Batch  262/538 - Train Accuracy: 0.9842, Validation Accuracy: 0.9549, Loss: 0.0222
Epoch   5 Batch  263/538 - Train Accuracy: 0.9709, Validation Accuracy: 0.9574, Loss: 0.0185
Epoch   5 Batch  264/538 - Train Accuracy: 0.9598, Validation Accuracy: 0.9593, Loss: 0.0226
Epoch   5 Batch  265/538 - Train Accuracy: 0.9686, Validation Accuracy: 0.9590, Loss: 0.0262
Epoch   5 Batch  266/538 - Train Accuracy: 0.9524, Validation Accuracy: 0.9622, Loss: 0.0226
Epoch   5 Batch  267/538 - Train Accuracy: 0.9654, Validation Accuracy: 0.9599, Loss: 0.0216
Epoch   5 Batch  268/538 - Train Accuracy: 0.9801, Validation Accuracy

Epoch   5 Batch  347/538 - Train Accuracy: 0.9691, Validation Accuracy: 0.9585, Loss: 0.0197
Epoch   5 Batch  348/538 - Train Accuracy: 0.9688, Validation Accuracy: 0.9663, Loss: 0.0162
Epoch   5 Batch  349/538 - Train Accuracy: 0.9820, Validation Accuracy: 0.9602, Loss: 0.0140
Epoch   5 Batch  350/538 - Train Accuracy: 0.9660, Validation Accuracy: 0.9579, Loss: 0.0251
Epoch   5 Batch  351/538 - Train Accuracy: 0.9586, Validation Accuracy: 0.9640, Loss: 0.0243
Epoch   5 Batch  352/538 - Train Accuracy: 0.9593, Validation Accuracy: 0.9600, Loss: 0.0308
Epoch   5 Batch  353/538 - Train Accuracy: 0.9711, Validation Accuracy: 0.9650, Loss: 0.0208
Epoch   5 Batch  354/538 - Train Accuracy: 0.9654, Validation Accuracy: 0.9615, Loss: 0.0186
Epoch   5 Batch  355/538 - Train Accuracy: 0.9818, Validation Accuracy: 0.9600, Loss: 0.0204
Epoch   5 Batch  356/538 - Train Accuracy: 0.9792, Validation Accuracy: 0.9574, Loss: 0.0145
Epoch   5 Batch  357/538 - Train Accuracy: 0.9730, Validation Accuracy

Epoch   5 Batch  436/538 - Train Accuracy: 0.9570, Validation Accuracy: 0.9725, Loss: 0.0229
Epoch   5 Batch  437/538 - Train Accuracy: 0.9848, Validation Accuracy: 0.9696, Loss: 0.0165
Epoch   5 Batch  438/538 - Train Accuracy: 0.9764, Validation Accuracy: 0.9675, Loss: 0.0157
Epoch   5 Batch  439/538 - Train Accuracy: 0.9883, Validation Accuracy: 0.9641, Loss: 0.0182
Epoch   5 Batch  440/538 - Train Accuracy: 0.9795, Validation Accuracy: 0.9631, Loss: 0.0184
Epoch   5 Batch  441/538 - Train Accuracy: 0.9699, Validation Accuracy: 0.9590, Loss: 0.0217
Epoch   5 Batch  442/538 - Train Accuracy: 0.9759, Validation Accuracy: 0.9558, Loss: 0.0156
Epoch   5 Batch  443/538 - Train Accuracy: 0.9544, Validation Accuracy: 0.9590, Loss: 0.0206
Epoch   5 Batch  444/538 - Train Accuracy: 0.9661, Validation Accuracy: 0.9588, Loss: 0.0186
Epoch   5 Batch  445/538 - Train Accuracy: 0.9830, Validation Accuracy: 0.9553, Loss: 0.0137
Epoch   5 Batch  446/538 - Train Accuracy: 0.9706, Validation Accuracy

Epoch   5 Batch  525/538 - Train Accuracy: 0.9766, Validation Accuracy: 0.9689, Loss: 0.0191
Epoch   5 Batch  526/538 - Train Accuracy: 0.9699, Validation Accuracy: 0.9641, Loss: 0.0187
Epoch   5 Batch  527/538 - Train Accuracy: 0.9732, Validation Accuracy: 0.9622, Loss: 0.0188
Epoch   5 Batch  528/538 - Train Accuracy: 0.9745, Validation Accuracy: 0.9684, Loss: 0.0202
Epoch   5 Batch  529/538 - Train Accuracy: 0.9646, Validation Accuracy: 0.9680, Loss: 0.0188
Epoch   5 Batch  530/538 - Train Accuracy: 0.9619, Validation Accuracy: 0.9680, Loss: 0.0233
Epoch   5 Batch  531/538 - Train Accuracy: 0.9666, Validation Accuracy: 0.9689, Loss: 0.0220
Epoch   5 Batch  532/538 - Train Accuracy: 0.9756, Validation Accuracy: 0.9700, Loss: 0.0167
Epoch   5 Batch  533/538 - Train Accuracy: 0.9799, Validation Accuracy: 0.9734, Loss: 0.0162
Epoch   5 Batch  534/538 - Train Accuracy: 0.9795, Validation Accuracy: 0.9775, Loss: 0.0127
Epoch   5 Batch  535/538 - Train Accuracy: 0.9725, Validation Accuracy

Epoch   6 Batch   78/538 - Train Accuracy: 0.9691, Validation Accuracy: 0.9627, Loss: 0.0182
Epoch   6 Batch   79/538 - Train Accuracy: 0.9710, Validation Accuracy: 0.9641, Loss: 0.0174
Epoch   6 Batch   80/538 - Train Accuracy: 0.9777, Validation Accuracy: 0.9570, Loss: 0.0147
Epoch   6 Batch   81/538 - Train Accuracy: 0.9762, Validation Accuracy: 0.9535, Loss: 0.0195
Epoch   6 Batch   82/538 - Train Accuracy: 0.9785, Validation Accuracy: 0.9629, Loss: 0.0170
Epoch   6 Batch   83/538 - Train Accuracy: 0.9629, Validation Accuracy: 0.9656, Loss: 0.0189
Epoch   6 Batch   84/538 - Train Accuracy: 0.9794, Validation Accuracy: 0.9613, Loss: 0.0182
Epoch   6 Batch   85/538 - Train Accuracy: 0.9830, Validation Accuracy: 0.9540, Loss: 0.0127
Epoch   6 Batch   86/538 - Train Accuracy: 0.9791, Validation Accuracy: 0.9494, Loss: 0.0152
Epoch   6 Batch   87/538 - Train Accuracy: 0.9652, Validation Accuracy: 0.9524, Loss: 0.0197
Epoch   6 Batch   88/538 - Train Accuracy: 0.9793, Validation Accuracy

Epoch   6 Batch  167/538 - Train Accuracy: 0.9702, Validation Accuracy: 0.9712, Loss: 0.0220
Epoch   6 Batch  168/538 - Train Accuracy: 0.9506, Validation Accuracy: 0.9686, Loss: 0.0217
Epoch   6 Batch  169/538 - Train Accuracy: 0.9947, Validation Accuracy: 0.9657, Loss: 0.0122
Epoch   6 Batch  170/538 - Train Accuracy: 0.9669, Validation Accuracy: 0.9668, Loss: 0.0180
Epoch   6 Batch  171/538 - Train Accuracy: 0.9719, Validation Accuracy: 0.9615, Loss: 0.0183
Epoch   6 Batch  172/538 - Train Accuracy: 0.9816, Validation Accuracy: 0.9625, Loss: 0.0131
Epoch   6 Batch  173/538 - Train Accuracy: 0.9818, Validation Accuracy: 0.9634, Loss: 0.0133
Epoch   6 Batch  174/538 - Train Accuracy: 0.9654, Validation Accuracy: 0.9524, Loss: 0.0176
Epoch   6 Batch  175/538 - Train Accuracy: 0.9855, Validation Accuracy: 0.9460, Loss: 0.0143
Epoch   6 Batch  176/538 - Train Accuracy: 0.9764, Validation Accuracy: 0.9600, Loss: 0.0196
Epoch   6 Batch  177/538 - Train Accuracy: 0.9840, Validation Accuracy

Epoch   6 Batch  256/538 - Train Accuracy: 0.9602, Validation Accuracy: 0.9673, Loss: 0.0171
Epoch   6 Batch  257/538 - Train Accuracy: 0.9913, Validation Accuracy: 0.9663, Loss: 0.0154
Epoch   6 Batch  258/538 - Train Accuracy: 0.9792, Validation Accuracy: 0.9656, Loss: 0.0157
Epoch   6 Batch  259/538 - Train Accuracy: 0.9853, Validation Accuracy: 0.9647, Loss: 0.0127
Epoch   6 Batch  260/538 - Train Accuracy: 0.9676, Validation Accuracy: 0.9574, Loss: 0.0166
Epoch   6 Batch  261/538 - Train Accuracy: 0.9795, Validation Accuracy: 0.9602, Loss: 0.0179
Epoch   6 Batch  262/538 - Train Accuracy: 0.9846, Validation Accuracy: 0.9600, Loss: 0.0183
Epoch   6 Batch  263/538 - Train Accuracy: 0.9693, Validation Accuracy: 0.9625, Loss: 0.0176
Epoch   6 Batch  264/538 - Train Accuracy: 0.9682, Validation Accuracy: 0.9625, Loss: 0.0183
Epoch   6 Batch  265/538 - Train Accuracy: 0.9621, Validation Accuracy: 0.9620, Loss: 0.0210
Epoch   6 Batch  266/538 - Train Accuracy: 0.9622, Validation Accuracy

Epoch   6 Batch  345/538 - Train Accuracy: 0.9868, Validation Accuracy: 0.9682, Loss: 0.0174
Epoch   6 Batch  346/538 - Train Accuracy: 0.9803, Validation Accuracy: 0.9643, Loss: 0.0185
Epoch   6 Batch  347/538 - Train Accuracy: 0.9785, Validation Accuracy: 0.9652, Loss: 0.0152
Epoch   6 Batch  348/538 - Train Accuracy: 0.9717, Validation Accuracy: 0.9663, Loss: 0.0135
Epoch   6 Batch  349/538 - Train Accuracy: 0.9896, Validation Accuracy: 0.9636, Loss: 0.0107
Epoch   6 Batch  350/538 - Train Accuracy: 0.9730, Validation Accuracy: 0.9657, Loss: 0.0209
Epoch   6 Batch  351/538 - Train Accuracy: 0.9729, Validation Accuracy: 0.9705, Loss: 0.0200
Epoch   6 Batch  352/538 - Train Accuracy: 0.9743, Validation Accuracy: 0.9734, Loss: 0.0279
Epoch   6 Batch  353/538 - Train Accuracy: 0.9758, Validation Accuracy: 0.9723, Loss: 0.0175
Epoch   6 Batch  354/538 - Train Accuracy: 0.9811, Validation Accuracy: 0.9709, Loss: 0.0138
Epoch   6 Batch  355/538 - Train Accuracy: 0.9848, Validation Accuracy

Epoch   6 Batch  434/538 - Train Accuracy: 0.9816, Validation Accuracy: 0.9657, Loss: 0.0142
Epoch   6 Batch  435/538 - Train Accuracy: 0.9648, Validation Accuracy: 0.9650, Loss: 0.0142
Epoch   6 Batch  436/538 - Train Accuracy: 0.9730, Validation Accuracy: 0.9673, Loss: 0.0205
Epoch   6 Batch  437/538 - Train Accuracy: 0.9867, Validation Accuracy: 0.9647, Loss: 0.0148
Epoch   6 Batch  438/538 - Train Accuracy: 0.9830, Validation Accuracy: 0.9703, Loss: 0.0112
Epoch   6 Batch  439/538 - Train Accuracy: 0.9945, Validation Accuracy: 0.9703, Loss: 0.0139
Epoch   6 Batch  440/538 - Train Accuracy: 0.9762, Validation Accuracy: 0.9659, Loss: 0.0160
Epoch   6 Batch  441/538 - Train Accuracy: 0.9807, Validation Accuracy: 0.9581, Loss: 0.0184
Epoch   6 Batch  442/538 - Train Accuracy: 0.9790, Validation Accuracy: 0.9613, Loss: 0.0138
Epoch   6 Batch  443/538 - Train Accuracy: 0.9693, Validation Accuracy: 0.9585, Loss: 0.0168
Epoch   6 Batch  444/538 - Train Accuracy: 0.9749, Validation Accuracy

Epoch   6 Batch  523/538 - Train Accuracy: 0.9855, Validation Accuracy: 0.9654, Loss: 0.0153
Epoch   6 Batch  524/538 - Train Accuracy: 0.9824, Validation Accuracy: 0.9664, Loss: 0.0145
Epoch   6 Batch  525/538 - Train Accuracy: 0.9846, Validation Accuracy: 0.9664, Loss: 0.0162
Epoch   6 Batch  526/538 - Train Accuracy: 0.9717, Validation Accuracy: 0.9664, Loss: 0.0153
Epoch   6 Batch  527/538 - Train Accuracy: 0.9828, Validation Accuracy: 0.9673, Loss: 0.0139
Epoch   6 Batch  528/538 - Train Accuracy: 0.9741, Validation Accuracy: 0.9700, Loss: 0.0186
Epoch   6 Batch  529/538 - Train Accuracy: 0.9670, Validation Accuracy: 0.9709, Loss: 0.0165
Epoch   6 Batch  530/538 - Train Accuracy: 0.9615, Validation Accuracy: 0.9705, Loss: 0.0184
Epoch   6 Batch  531/538 - Train Accuracy: 0.9785, Validation Accuracy: 0.9728, Loss: 0.0164
Epoch   6 Batch  532/538 - Train Accuracy: 0.9771, Validation Accuracy: 0.9727, Loss: 0.0145
Epoch   6 Batch  533/538 - Train Accuracy: 0.9861, Validation Accuracy

Epoch   7 Batch   76/538 - Train Accuracy: 0.9740, Validation Accuracy: 0.9622, Loss: 0.0182
Epoch   7 Batch   77/538 - Train Accuracy: 0.9785, Validation Accuracy: 0.9608, Loss: 0.0133
Epoch   7 Batch   78/538 - Train Accuracy: 0.9771, Validation Accuracy: 0.9668, Loss: 0.0154
Epoch   7 Batch   79/538 - Train Accuracy: 0.9775, Validation Accuracy: 0.9693, Loss: 0.0157
Epoch   7 Batch   80/538 - Train Accuracy: 0.9812, Validation Accuracy: 0.9636, Loss: 0.0139
Epoch   7 Batch   81/538 - Train Accuracy: 0.9744, Validation Accuracy: 0.9682, Loss: 0.0192
Epoch   7 Batch   82/538 - Train Accuracy: 0.9881, Validation Accuracy: 0.9663, Loss: 0.0173
Epoch   7 Batch   83/538 - Train Accuracy: 0.9629, Validation Accuracy: 0.9636, Loss: 0.0168
Epoch   7 Batch   84/538 - Train Accuracy: 0.9749, Validation Accuracy: 0.9643, Loss: 0.0165
Epoch   7 Batch   85/538 - Train Accuracy: 0.9929, Validation Accuracy: 0.9567, Loss: 0.0130
Epoch   7 Batch   86/538 - Train Accuracy: 0.9871, Validation Accuracy

Epoch   7 Batch  165/538 - Train Accuracy: 0.9805, Validation Accuracy: 0.9719, Loss: 0.0119
Epoch   7 Batch  166/538 - Train Accuracy: 0.9869, Validation Accuracy: 0.9691, Loss: 0.0132
Epoch   7 Batch  167/538 - Train Accuracy: 0.9753, Validation Accuracy: 0.9691, Loss: 0.0177
Epoch   7 Batch  168/538 - Train Accuracy: 0.9723, Validation Accuracy: 0.9767, Loss: 0.0183
Epoch   7 Batch  169/538 - Train Accuracy: 0.9957, Validation Accuracy: 0.9743, Loss: 0.0104
Epoch   7 Batch  170/538 - Train Accuracy: 0.9736, Validation Accuracy: 0.9700, Loss: 0.0145
Epoch   7 Batch  171/538 - Train Accuracy: 0.9775, Validation Accuracy: 0.9695, Loss: 0.0148
Epoch   7 Batch  172/538 - Train Accuracy: 0.9754, Validation Accuracy: 0.9693, Loss: 0.0135
Epoch   7 Batch  173/538 - Train Accuracy: 0.9860, Validation Accuracy: 0.9627, Loss: 0.0107
Epoch   7 Batch  174/538 - Train Accuracy: 0.9754, Validation Accuracy: 0.9599, Loss: 0.0139
Epoch   7 Batch  175/538 - Train Accuracy: 0.9834, Validation Accuracy

Epoch   7 Batch  254/538 - Train Accuracy: 0.9645, Validation Accuracy: 0.9636, Loss: 0.0177
Epoch   7 Batch  255/538 - Train Accuracy: 0.9783, Validation Accuracy: 0.9611, Loss: 0.0118
Epoch   7 Batch  256/538 - Train Accuracy: 0.9717, Validation Accuracy: 0.9668, Loss: 0.0144
Epoch   7 Batch  257/538 - Train Accuracy: 0.9907, Validation Accuracy: 0.9693, Loss: 0.0112
Epoch   7 Batch  258/538 - Train Accuracy: 0.9771, Validation Accuracy: 0.9643, Loss: 0.0132
Epoch   7 Batch  259/538 - Train Accuracy: 0.9929, Validation Accuracy: 0.9663, Loss: 0.0101
Epoch   7 Batch  260/538 - Train Accuracy: 0.9695, Validation Accuracy: 0.9615, Loss: 0.0144
Epoch   7 Batch  261/538 - Train Accuracy: 0.9811, Validation Accuracy: 0.9556, Loss: 0.0136
Epoch   7 Batch  262/538 - Train Accuracy: 0.9854, Validation Accuracy: 0.9592, Loss: 0.0156
Epoch   7 Batch  263/538 - Train Accuracy: 0.9744, Validation Accuracy: 0.9632, Loss: 0.0135
Epoch   7 Batch  264/538 - Train Accuracy: 0.9672, Validation Accuracy

Epoch   7 Batch  343/538 - Train Accuracy: 0.9895, Validation Accuracy: 0.9783, Loss: 0.0110
Epoch   7 Batch  344/538 - Train Accuracy: 0.9742, Validation Accuracy: 0.9771, Loss: 0.0115
Epoch   7 Batch  345/538 - Train Accuracy: 0.9855, Validation Accuracy: 0.9771, Loss: 0.0136
Epoch   7 Batch  346/538 - Train Accuracy: 0.9781, Validation Accuracy: 0.9721, Loss: 0.0162
Epoch   7 Batch  347/538 - Train Accuracy: 0.9811, Validation Accuracy: 0.9737, Loss: 0.0122
Epoch   7 Batch  348/538 - Train Accuracy: 0.9881, Validation Accuracy: 0.9712, Loss: 0.0100
Epoch   7 Batch  349/538 - Train Accuracy: 0.9859, Validation Accuracy: 0.9737, Loss: 0.0089
Epoch   7 Batch  350/538 - Train Accuracy: 0.9808, Validation Accuracy: 0.9700, Loss: 0.0173
Epoch   7 Batch  351/538 - Train Accuracy: 0.9736, Validation Accuracy: 0.9672, Loss: 0.0140
Epoch   7 Batch  352/538 - Train Accuracy: 0.9695, Validation Accuracy: 0.9714, Loss: 0.0227
Epoch   7 Batch  353/538 - Train Accuracy: 0.9760, Validation Accuracy

Epoch   7 Batch  432/538 - Train Accuracy: 0.9792, Validation Accuracy: 0.9558, Loss: 0.0132
Epoch   7 Batch  433/538 - Train Accuracy: 0.9771, Validation Accuracy: 0.9572, Loss: 0.0258
Epoch   7 Batch  434/538 - Train Accuracy: 0.9668, Validation Accuracy: 0.9597, Loss: 0.0106
Epoch   7 Batch  435/538 - Train Accuracy: 0.9658, Validation Accuracy: 0.9629, Loss: 0.0138
Epoch   7 Batch  436/538 - Train Accuracy: 0.9773, Validation Accuracy: 0.9663, Loss: 0.0146
Epoch   7 Batch  437/538 - Train Accuracy: 0.9861, Validation Accuracy: 0.9650, Loss: 0.0102
Epoch   7 Batch  438/538 - Train Accuracy: 0.9826, Validation Accuracy: 0.9636, Loss: 0.0104
Epoch   7 Batch  439/538 - Train Accuracy: 0.9889, Validation Accuracy: 0.9600, Loss: 0.0146
Epoch   7 Batch  440/538 - Train Accuracy: 0.9871, Validation Accuracy: 0.9592, Loss: 0.0132
Epoch   7 Batch  441/538 - Train Accuracy: 0.9750, Validation Accuracy: 0.9604, Loss: 0.0143
Epoch   7 Batch  442/538 - Train Accuracy: 0.9860, Validation Accuracy

Epoch   7 Batch  521/538 - Train Accuracy: 0.9830, Validation Accuracy: 0.9680, Loss: 0.0140
Epoch   7 Batch  522/538 - Train Accuracy: 0.9768, Validation Accuracy: 0.9689, Loss: 0.0109
Epoch   7 Batch  523/538 - Train Accuracy: 0.9814, Validation Accuracy: 0.9721, Loss: 0.0141
Epoch   7 Batch  524/538 - Train Accuracy: 0.9848, Validation Accuracy: 0.9696, Loss: 0.0109
Epoch   7 Batch  525/538 - Train Accuracy: 0.9840, Validation Accuracy: 0.9695, Loss: 0.0128
Epoch   7 Batch  526/538 - Train Accuracy: 0.9779, Validation Accuracy: 0.9680, Loss: 0.0140
Epoch   7 Batch  527/538 - Train Accuracy: 0.9818, Validation Accuracy: 0.9588, Loss: 0.0120
Epoch   7 Batch  528/538 - Train Accuracy: 0.9774, Validation Accuracy: 0.9620, Loss: 0.0163
Epoch   7 Batch  529/538 - Train Accuracy: 0.9742, Validation Accuracy: 0.9695, Loss: 0.0149
Epoch   7 Batch  530/538 - Train Accuracy: 0.9752, Validation Accuracy: 0.9737, Loss: 0.0140
Epoch   7 Batch  531/538 - Train Accuracy: 0.9803, Validation Accuracy

### Save Parameters
Save the `batch_size` and `save_path` parameters for inference.

In [29]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
# Save parameters for checkpoint
helper.save_params(save_path)

# Checkpoint

In [30]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
import tensorflow as tf
import numpy as np
import helper
import problem_unittests as tests

_, (source_vocab_to_int, target_vocab_to_int), (source_int_to_vocab, target_int_to_vocab) = helper.load_preprocess()
load_path = helper.load_params()

## Sentence to Sequence
To feed a sentence into the model for translation, you first need to preprocess it.  Implement the function `sentence_to_seq()` to preprocess new sentences.

- Convert the sentence to lowercase
- Convert words into ids using `vocab_to_int`
 - Convert words not in the vocabulary, to the `<UNK>` word id.

In [31]:
def sentence_to_seq(sentence, vocab_to_int):
    """
    Convert a sentence to a sequence of ids
    :param sentence: String
    :param vocab_to_int: Dictionary to go from the words to an id
    :return: List of word ids
    """
    id_list = []
    for word in sentence.split():
        id_list.append(vocab_to_int.get(word.lower(), vocab_to_int['<UNK>']))
        
    return id_list


"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_sentence_to_seq(sentence_to_seq)

Tests Passed


## Translate
This will translate `translate_sentence` from English to French.

In [33]:
translate_sentence = 'he saw a old yellow truck .'
batch_size=256


"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
translate_sentence = sentence_to_seq(translate_sentence, source_vocab_to_int)

loaded_graph = tf.Graph()
with tf.Session(graph=loaded_graph) as sess:
    # Load saved model
    loader = tf.train.import_meta_graph(load_path + '.meta')
    loader.restore(sess, load_path)

    input_data = loaded_graph.get_tensor_by_name('input:0')
    logits = loaded_graph.get_tensor_by_name('predictions:0')
    target_sequence_length = loaded_graph.get_tensor_by_name('target_sequence_length:0')
    source_sequence_length = loaded_graph.get_tensor_by_name('source_sequence_length:0')
    keep_prob = loaded_graph.get_tensor_by_name('keep_prob:0')

    translate_logits = sess.run(logits, {input_data: [translate_sentence]*batch_size,
                                         target_sequence_length: [len(translate_sentence)*2]*batch_size,
                                         source_sequence_length: [len(translate_sentence)]*batch_size,
                                         keep_prob: 1.0})[0]

print('Input')
print('  Word Ids:      {}'.format([i for i in translate_sentence]))
print('  English Words: {}'.format([source_int_to_vocab[i] for i in translate_sentence]))

print('\nPrediction')
print('  Word Ids:      {}'.format([i for i in translate_logits]))
print('  French Words: {}'.format(" ".join([target_int_to_vocab[i] for i in translate_logits])))


INFO:tensorflow:Restoring parameters from checkpoints/dev
Input
  Word Ids:      [136, 166, 225, 186, 157, 25, 164]
  English Words: ['he', 'saw', 'a', 'old', 'yellow', 'truck', '.']

Prediction
  Word Ids:      [44, 169, 75, 264, 346, 60, 333, 1]
  French Words: il a vu rouillé camion jaune . <EOS>


## Imperfect Translation
You might notice that some sentences translate better than others.  Since the dataset you're using only has a vocabulary of 227 English words of the thousands that you use, you're only going to see good results using these words.  For this project, you don't need a perfect translation. However, if you want to create a better translation model, you'll need better data.

You can train on the [WMT10 French-English corpus](http://www.statmt.org/wmt10/training-giga-fren.tar).  This dataset has more vocabulary and richer in topics discussed.  However, this will take you days to train, so make sure you've a GPU and the neural network is performing well on dataset we provided.  Just make sure you play with the WMT10 corpus after you've submitted this project.
## Submitting This Project
When submitting this project, make sure to run all the cells before saving the notebook. Save the notebook file as "dlnd_language_translation.ipynb" and save it as a HTML file under "File" -> "Download as". Include the "helper.py" and "problem_unittests.py" files in your submission.